### PR TITLE
added objectstore variable available storage

### PR DESCRIPTION
**Pull request by voroyam**
new variable was added for the metrics app to fix 
https://github.com/owncloud/enterprise/issues/5384

docs 
https://doc.owncloud.com/server/next/admin_manual/configuration/files/external_storage/s3_compatible_object_storage_as_primary.html#amazon-s3

pr 
https://github.com/owncloud/metrics/pull/159

### CODE REVIEW

**Review by TheRealBro — CHANGES_REQUESTED** [review]
I understand the need for the new variable. But the documentation states 1 TB as example value. Why should we always set this value to 1TB by default in our base image? wouldn't that break other things at some points?

**Comment by voroyam** [comment]
> wouldn't that break other things at some points?

The value of the variable is just for cosmetics. It shows the total amount of storage, based on that you see a pie chart of used storage vs free storage. As far as I know nothing else depends on that value, for example no quotas. We can leave that value empty. I just need to have that variable in the image so I can use it for the bayernbox image that is based on that

**Review by xoxys — CHANGES_REQUESTED** [review]
It's just wrong to set a cosmetic value by default for everyone, with a random default. Please set an empty default and make it conditional in the template as we do in e.g. https://github.com/owncloud-docker/base/blob/master/v20.04/overlay/etc/templates/config.php#L4

**Comment by voroyam** [comment]
> It's just wrong to set a cosmetic value by default for everyone, with a random default. Please set an empty default and make it conditional in the template as we do in e.g. https://github.com/owncloud-docker/base/blob/master/v20.04/overlay/etc/templates/config.php#L4

conditional to what?

**Comment by TheRealBro** [comment]
> > It's just wrong to set a cosmetic value by default for everyone, with a random default. Please set an empty default and make it conditional in the template as we do in e.g. https://github.com/owncloud-docker/base/blob/master/v20.04/overlay/etc/templates/config.php#L4
> 
> conditional to what?

Conditional, as in the linked example by @xoxys -> if empty -> not set, otherwise set, if it contains a value.

**Comment by voroyam** [comment]
like this? 

```
if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
    $availableStorage = array_map('trim', explode(',', getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE')));
  } else {
    $availableStorage = [explode(':', getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'))[0]];
  }
```

**Comment by TheRealBro** [comment]
> or this?
> 
> ```
> if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
>     $availableStorage = (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'));
>   }
> ```

Like this ^^ would be the way to go. And in deployment, the value can be set or not set.

**Comment by voroyam** [comment]
> > or this?
> > ```
> > if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
> >     $availableStorage = (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'));
> >   }
> > ```
> 
> Like this ^^ would be the way to go. And in deployment, the value can be set or not set.

okay, full thing is this? 

```
<?php

function getObjectstoreFromEnv() {
  if (getenv('OWNCLOUD_OBJECTSTORE_ENABLED') && getenv('OWNCLOUD_OBJECTSTORE_ENABLED') === 'true') {
    return [
      'objectstore' => [
        'class' => getenv('OWNCLOUD_OBJECTSTORE_CLASS'),
        'arguments' => [
          'bucket' => getenv('OWNCLOUD_OBJECTSTORE_BUCKET'),
            if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
              $availableStorage = (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'));
            }
          'availableStorage' => getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'),
          'part_size' => getenv('OWNCLOUD_OBJECTSTORE_PART_SIZE'),
          'concurrency' => getenv('OWNCLOUD_OBJECTSTORE_CONCURRENCY'),
          'options' => [
            'endpoint' => getenv('OWNCLOUD_OBJECTSTORE_ENDPOINT'),
            'version' => getenv('OWNCLOUD_OBJECTSTORE_VERSION'),
            'region' => getenv('OWNCLOUD_OBJECTSTORE_REGION'),
            'use_path_style_endpoint' => getenv('OWNCLOUD_OBJECTSTORE_PATHSTYLE') === 'true',
            'command.params' => [
              'PathStyle' => getenv('OWNCLOUD_OBJECTSTORE_PATHSTYLE') === 'true',
            ],
            'credentials' => [
              'key'   => getenv('OWNCLOUD_OBJECTSTORE_KEY'),
              'secret'  => getenv('OWNCLOUD_OBJECTSTORE_SECRET'),
            ],
          ],
        ],
      ],
    ];
  }

  return [];
}

$CONFIG = getObjectstoreFromEnv();
```

File: v20.04/overlay/etc/templates/objectstore.php.tmpl [hunk 1]
```diff
@@ -7,6 +7,7 @@ function getObjectstoreFromEnv() {
         'class' => getenv('OWNCLOUD_OBJECTSTORE_CLASS'),
         'arguments' => [
           'bucket' => getenv('OWNCLOUD_OBJECTSTORE_BUCKET'),
+          'availableStorage' => getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'),
```
Comment (xoxys) [inline v20.04/overlay/etc/templates/objectstore.php.tmpl:10]:
```suggestion
          if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
          'availableStorage' => getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'),
          }
```

**Review by xoxys — CHANGES_REQUESTED** [review]
No, why do you add a dollar now? This was just an example, of course it needs to be adjusted to _this_ particular_  config file and syntax...

File: v20.04/overlay/etc/templates/objectstore.php.tmpl (hunk 1 above)
Comment (TheRealBro) [reply v20.04/overlay/etc/templates/objectstore.php.tmpl:10]: HALLELUJAH!

**Comment by xoxys** [comment]
However, this also might not work as this is part of a return statement... Needs to be tested locally first before it could be merged.

File: v20.04/overlay/etc/templates/objectstore.php.tmpl (hunk 1 above)
Comment (voroyam) [reply v20.04/overlay/etc/templates/objectstore.php.tmpl:10]:
```          
'bucket' => getenv('OWNCLOUD_OBJECTSTORE_BUCKET'),
          if (getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE') != '') {
          'availableStorage' => getenv('OWNCLOUD_OBJECTSTORE_AVAILABLE_STORAGE'),
          },
          'part_size' => getenv('OWNCLOUD_OBJECTSTORE_PART_SIZE'),
```

**Comment by voroyam** [comment]
does not seem to work:

```
"objectstore": {
            "class": "OCA\\Files_Primary_S3\\S3Storage",
            "arguments": {
                "bucket": "owncloud",
                "part_size": "268435456",
                "concurrency": "3",
                "options": {
                    "endpoint": "http:\/\/scality:8000",
                    "version": "2006-03-01",
                    "region": "us-east-1",
                    "use_path_style_endpoint": true,
                    "command.params": {
                        "PathStyle": true
                    },
                    "credentials": {
                        "key": "***REMOVED SENSITIVE VALUE***",
                        "secret": "***REMOVED SENSITIVE VALUE***"
                    }
                }
            }
        },
```

**Comment by voroyam** [comment]
Any idea what needs to be changed to get it to work?

**Comment by voroyam** [comment]
> Inline conditionals within an array definition don't work.

if I put your suggestion in the container goes in to a loop and breaks

**Comment by xoxys** [comment]
Works for me ... What exactly have you tested? 

```
docker exec -it owncloud_server occ config:system:get objectstore
class: OCA\Files_Primary_S3\S3Storage
arguments:
  bucket: owncloud
  part_size: 5242880
  concurrency: 3
  options:
    endpoint: http://minio:9000
    version: 2006-03-01
    region: us-east-1
    use_path_style_endpoint: true
    command.params:
      PathStyle: true
    credentials:
      key: minioadmin
      secret: minioadmin
  availableStorage: 12345678900000
```

**Review by xoxys — APPROVED** [review]

**Pull request merged**